File: 3742-maximum-path-score-in-a-grid/3742-maximum-path-score-in-a-grid.java
class Solution {
    public int maxPathScore(int[][] grid, int K) {
        if (minPathSum(grid) > K) {
            return -1;
        }
        int m = grid.length;
        int n = grid[0].length;
        K = Math.min(K, m + n - 2);
        int[][] f = new int[n + 1][K + 2];
        for (int[] row : f) {
            Arrays.fill(row, Integer.MIN_VALUE);
        }
        f[1][1] = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                int x = grid[i][j];
                for (int k = Math.min(K, i + j); k >= 0; k--) { 
                    int newK = x > 0 ? k - 1 : k;
                    f[j + 1][k + 1] = Math.max(f[j + 1][newK + 1], f[j][newK + 1]) + x;
                }
            }
        }
        int ans = 0;
        for (int x : f[n]) {
            ans = Math.max(ans, x);
        }
        return ans;
    }
    private int minPathSum(int[][] grid) {
        int n = grid[0].length;
        int[] f = new int[n + 1];
        Arrays.fill(f, Integer.MAX_VALUE);
        f[1] = 0;
        for (int[] row : grid) {
            for (int j = 0; j < n; j++) {
                f[j + 1] = Math.min(f[j], f[j + 1]) + Math.min(row[j], 1); 
            }
        }
        return f[n];
    }
}

File: 3742-maximum-path-score-in-a-grid/3742-maximum-path-score-in-a-grid.cpp
short dp[200][200][1001];
class Solution {
public:
    int k;
    const short NO=SHRT_MIN;
    int f(int i, int j, int c, vector<vector<int>>& grid){
        if (i<0 || j<0 || c>k) return NO;
        if (dp[i][j][c]!=-1) return dp[i][j][c];
        int x=grid[i][j], cost=c+(x!=0);
        if (cost>k) 
            return dp[i][j][c]=NO;
        if (i==0 && j==0)
            return dp[i][j][c]=grid[0][0];
        return dp[i][j][c]=x+max(f(i-1, j, cost, grid), f(i, j-1, cost, grid));
    }
    int maxPathScore(vector<vector<int>>& grid, int k) {
        this->k=k;
        const int row=grid.size(), col=grid[0].size();
        for(int i=0; i<row; i++)
            for(int j=0; j<col; j++)
                memset(dp[i][j], -1, (k+1)*sizeof(short));
        int ans=f(row-1, col-1, 0, grid);
        return ans<0?-1:ans;
    }
};

const auto init = []() noexcept {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 'c';
}();